### ISSUE TITLE
POINT shapefiles require dataset extent to be drawn in 2.0.0-dev

### ISSUE THREAD

**Issue by pstatix** [GeospatialPython/pyshp#130]
When creating a POINT shapefile, the `__shapefileHeader()` method is supplying the file header with a unpacked `*self.bbox()` of `0, 0, 0, 0`. What this results in is geometry that has no spatial index. This occurs due to the `self.__bbox()` method never getting called for shapes of type `1` in the `__shpRecord()` method. The `Writer` instance has a `self._bbox` initialized to `[0, 0, 0, 0]` and without calling it ever again, the `save()` method will ultimately try and pack the extent of a point in the file header with a bounding box as `0, 0, 0, 0`. ArcGIS is unable to plot points since no extent is provided, only when another file with an extent is loaded will the points display.

To resolve this (temporarily on a local copy) I've done the following:

    def point(self, x, y, z=0, m=0, shapeType=POINT):
        """Creates a point shape."""
        pointShape = Shape(shapeType)
        if shapeType == POINT:
            pointShape.points.append([x, y])
        elif shapeType == POINTZ:
            pointShape.points.append([x, y, z])
        elif shapeType == POINTM:
            pointShape.points.append([x, y, z, m])
        self.shape(pointShape)
        _ = self.__bbox(pointShape) # catch a throw-away return

All it does is force the global `self._bbox` to update to the extent of the provided points. Since the new implementation of the `self.__bbox()` method updates by comparing the initial `self._bbox()` to the supplied point, the minimum x & y will always be 0 0. For example, if I added a `print()` inside the `__shpRecord()` method when packing the bounding box:

    w = shapefile.Writer(1, True)
    w.field('Point')
    w.record('First')
    w.record('Second')
    w.point(1.0, 2.0)
    w.point(3.0, 4.0)
    w.save('/some/path')
    >>> 0 0 3.0 4.0 # when the 'saveShp()' method calls
    >>> 0 0 3.0 4.0 # when the 'saveShx()' method calls

So in summary, a redesign of how the `self.__bbox()` method is called when dealing with POINTs needs to be performed so that proper extent is written.

**Comment by karimbahgat** [comment]
Initiating bbox with 0s was indeed problematic. Instead, initiate with None and explicitly test if the bbox is None or is actually valid, and some other changes. This fixes the bug with wrong bbox for single point shapefiles. See 2782be8. Tested that bbox now works on a varity of cases including this.
